src/main/webapp/app/exercises/programming/manage/grading/charts/test-case-distribution-chart.component.ts

```diff
@@ -74,50 +143,73 @@ export class TestCaseDistributionChartComponent implements OnChanges {
 
                 const label = element.label;
-                const backgroundColor = this.getColor(+i / this.testCases.length, 50);
-                const hoverBackgroundColor = this.getColor(+i / this.testCases.length, 60);
-
-                testCaseColors[label] = backgroundColor;
-
-                this.weightChartDatasets.push({
-                    label,
-                    backgroundColor,
-                    hoverBackgroundColor,
-                    data: [element.relWeight, element.relScore],
-                });
-
-                this.pointsChartDatasets.push({
-                    label,
-                    backgroundColor,
-                    hoverBackgroundColor,
-                    data: [element.relPoints],
-                });
+                const color = this.getColor(+i / this.testCases.length, 50);
+
+                weight.series.push({ name: element.label, value: element.relWeight, bonus: element.relScore });
+                weightAndBonus.series.push({ name: element.label, value: element.relScore, weight: element.relWeight });
+
+                points.series.push({ name: element.label, value: element.relPoints });
+
+                testCaseColors[label] = color;
+                this.ngxColors.domain.push(color);
             }
 
+            this.ngxWeightData.push(weight);
+            this.ngxWeightData.push(weightAndBonus);
+
+            this.ngxPointsData.push(points);
+
             // update colors for test case table
             this.testCaseColorsChange.emit(testCaseColors);
         } else {
             // update values in-place
             for (let i = 0; i < testCaseScores.length; i++) {
                 const element = testCaseScores[i];
-                this.weightChartDatasets[i].data![0] = element.relWeight;
-                this.weightChartDatasets[i].data![1] = element.relScore;
-                this.pointsChartDatasets[i].data![0] = element.relPoints;
+
+                this.ngxWeightData[0].series[i].value = element.relWeight;
+                this.ngxWeightData[1].series[i].value = element.relScore;
+                this.ngxPointsData[0].series[i].value = element.relPoints;
             }
         }
+
+        this.ngxWeightData = [...this.ngxWeightData];
+        this.ngxPointsData = [...this.ngxPointsData];
     }
 
+    /**
+     * Dynamically generates a color based on the input
+     * @param i factor that is modifying the first coordinate of the color
+     * @param l percentage defining the last part of the color
+     * @returns color in hsl format
+     */
     getColor(i: number, l: number): string {
         return `hsl(${(i * 360 * 3) % 360}, 55%, ${l}%)`;
     }
+
+    /**
+     * Appends a percentage sign to every tick on the x axis
+     * @param tick the default tick label as string
+     * @returns tick label string that is extended by an percentage sign
+     */
+    xAxisFormatting(tick: string): string {
+        return tick + '%';
+    }
```

Comment:
This is already identically in `sca-category-distribution-chart.component.ts`. Can it be moved to a utility file to reduce duplication?

Reply:
I introduced a utils in f29991a